### PR TITLE
Standardize time units per REP-103

### PR DESCRIPTION
**Pull request by simutisernestas**
Addressing #2096.

Is "migration guide updates" referring to https://navigation.ros.org/migration/Foxy.html?

### CODE REVIEW

**Comment by SteveMacenski** [comment]
@simutisernestas that is correct - its in the navigation.ros.org repo on the ros-planning org.

File: doc/parameters/param_list.md [hunk 1]
```diff
@@ -429,15 +429,15 @@ When `controller_plugins`\`progress_checker_plugin`\`goal_checker_plugin` parame
 | ----------| --------| ------------|
 | node_names | N/A | Ordered list of node names to bringup through lifecycle transition |
 | autostart | false | Whether to transition nodes to active state on startup |
-| bond_timeout_ms | 4000 | Timeout for bond to fail if no heartbeat can be found, in milliseconds. If set to 0, it will be disabled. Must be larger than 300ms for stable bringup. |
+| bond_timeout | 4 | Timeout for bond to fail if no heartbeat can be found, in milliseconds. If set to 0, it will be disabled. Must be larger than 0.3s for stable bringup. |
```
Comment (SteveMacenski) [inline doc/parameters/param_list.md:432]: This should be a double now, so 4.0 seconds, so that people can still represent milliseconds. Also, the description needs to be updated (no longer ms)

File: doc/parameters/param_list.md [hunk 2]
```diff
@@ -429,15 +429,15 @@ When `controller_plugins`\`progress_checker_plugin`\`goal_checker_plugin` parame
...
 
 # map_server
 
 ## map_saver
 
 | Parameter | Default | Description |
 | ----------| --------| ------------|
-| save_map_timeout | 2000 | Timeout to attempt to save map with (ms) |
+| save_map_timeout | 2 | Timeout to attempt to save map with (s) |
```
Comment (SteveMacenski) [inline doc/parameters/param_list.md:440]: Double now as well

File: nav2_bringup/bringup/params/nav2_params.yaml [hunk 3]
```diff
@@ -240,7 +240,7 @@ map_server:
 map_saver:
   ros__parameters:
     use_sim_time: True
-    save_map_timeout: 5000
+    save_map_timeout: 5
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params.yaml:243]: `5.0`

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 4]
```diff
@@ -40,13 +40,13 @@ LifecycleManager::LifecycleManager()
   // of nodes
   declare_parameter("node_names");
   declare_parameter("autostart", rclcpp::ParameterValue(false));
-  declare_parameter("bond_timeout_ms", 4000);
+  declare_parameter("bond_timeout", 4);
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:43]: `4.0`

File: nav2_map_server/launch/map_saver_server.launch.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ def generate_launch_description():
     lifecycle_nodes = ['map_saver']
     use_sim_time = True
     autostart = True
-    save_map_timeout = 2000
+    save_map_timeout = 2
```
Comment (SteveMacenski) [inline nav2_map_server/launch/map_saver_server.launch.py:26]: `2.0`

File: nav2_map_server/src/map_saver/map_saver.cpp [hunk 6]
```diff
@@ -47,7 +47,7 @@ MapSaver::MapSaver()
   RCLCPP_INFO(get_logger(), "Creating");
 
   save_map_timeout_ = std::make_shared<rclcpp::Duration>(
-    std::chrono::milliseconds(declare_parameter("save_map_timeout", 2000)));
+    std::chrono::seconds(declare_parameter("save_map_timeout", 2)));
```
Comment (SteveMacenski) [inline nav2_map_server/src/map_saver/map_saver.cpp:50]: `2.0`

File: nav2_map_server/test/map_saver_params.yaml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 map_saver:
     ros__parameters:
-        save_map_timeout: 5000
+        save_map_timeout: 5
```
Comment (SteveMacenski) [inline nav2_map_server/test/map_saver_params.yaml:3]: `5.0`

File: nav2_system_tests/src/costmap_filters/keepout_params.yaml [hunk 8]
```diff
@@ -248,7 +248,7 @@ map_server:
 map_saver:
   ros__parameters:
     use_sim_time: True
-    save_map_timeout: 5000
+    save_map_timeout: 5
```
Comment (SteveMacenski) [inline nav2_system_tests/src/costmap_filters/keepout_params.yaml:251]: `5.0`

File: nav2_smac_planner/src/smac_planner_2d.cpp [hunk 9]
```diff
@@ -88,8 +88,8 @@ void SmacPlanner2D::configure(
   node->get_parameter(name + ".minimum_turning_radius", minimum_turning_radius);
 
   nav2_util::declare_parameter_if_not_declared(
-    node, name + ".max_planning_time_ms", rclcpp::ParameterValue(1000.0));
-  node->get_parameter(name + ".max_planning_time_ms", _max_planning_time);
+    node, name + ".max_planning_time", rclcpp::ParameterValue(1000.0));
```
Comment (SteveMacenski) [inline nav2_smac_planner/src/smac_planner_2d.cpp:91]: `1.0`

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Big thing is making sure to use the right types. ROS2 is really picky, if you define a default of type `double`, then it will crash if you try to give it a `int`. These should all be doubles now as seconds.

File: doc/parameters/param_list.md [hunk 10]
```diff
@@ -429,15 +429,15 @@ When `controller_plugins`\`progress_checker_plugin`\`goal_checker_plugin` parame
 | ----------| --------| ------------|
```
Comment (SteveMacenski) [inline doc/parameters/param_list.md:429]: All the changes to this `md` file have analog updates that need to be made to navigation.ros.org under the configuration section. Each package has a page with parameter descriptions like this.

File: nav2_bringup/bringup/params/nav2_params.yaml [hunk 11]
```diff
@@ -240,7 +240,7 @@ map_server:
 map_saver:
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params.yaml:240]: There are 2 more yaml files in this directory that need to be updated

File: nav2_smac_planner/src/smac_planner.cpp [hunk 12]
```diff
@@ -111,8 +111,8 @@ void SmacPlanner::configure(
   node->get_parameter(name + ".analytic_expansion_ratio", search_info.analytic_expansion_ratio);
 
   nav2_util::declare_parameter_if_not_declared(
-    node, name + ".max_planning_time_ms", rclcpp::ParameterValue(5000.0));
-  node->get_parameter(name + ".max_planning_time_ms", _max_planning_time);
+    node, name + ".max_planning_time", rclcpp::ParameterValue(5.0));
```
Comment (SteveMacenski) [inline nav2_smac_planner/src/smac_planner.cpp:114]: I see you changed the units of the parameter, but you don't make any updates in the code itself where it is assumed to be in `ms` right now?

Comment (simutisernestas) [reply nav2_smac_planner/src/smac_planner.cpp:114]: The only place it's used is [here](https://github.com/ros-planning/navigation2/blob/3bbb4fe020194f2d866ebe50a2ae82e43e260655/nav2_smac_planner/src/smac_planner.cpp#L328). I've tested similar peace of code in a sandbox: 
```#include <iostream>
#include <chrono>
#include <thread>

using namespace std::chrono_literals;
using namespace std::chrono;

int main(int argc, char *argv[])
{
	double _max_planning_time = 5.0;
	steady_clock::time_point a = steady_clock::now();
	std::this_thread::sleep_for(100ms);
	steady_clock::time_point b = steady_clock::now();
	duration<double> time_span = duration_cast<duration<double>>(b - a);
	double time_remaining = _max_planning_time - static_cast<double>(time_span.count()); // TODO
	std::cout << time_remaining << '\n';
	std::cout << static_cast<double>(time_span.count()) << '\n';
	return 0;
}
```
Output:
```
4.89993
0.100073
```
That ```time_span``` variable is already in seconds. I guess the previous implementation was flawed actually.

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 13]
```diff
@@ -146,19 +144,17 @@ LifecycleManager::destroyLifecycleServiceClients()
 bool
 LifecycleManager::createBondConnection(const std::string & node_name)
 {
-  const double timeout_ns =
-    std::chrono::duration_cast<std::chrono::nanoseconds>(bond_timeout_).count();
-  const double timeout_s = timeout_ns / 1e9;
+  const double timeout_s = bond_timeout_.count();
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:147]: My understanding is that `count` gives you back out the count of the units, which in this case is now seconds. This would then lose all sig figs below an integer, which could mess up the other things in this function. I think you should revert `bond_timeout_` to a `chrono milliseconds` or `nanoseconds` variable type and convert the `s`-typed parameter into it and update this function appropriately.

I think `timeout_s` would be 0 in alot of cases and the `from_seconds` cast I'm not sure what it will do.

File: nav2_smac_planner/src/smac_planner.cpp (hunk 12 above)
Comment (SteveMacenski) [reply nav2_smac_planner/src/smac_planner.cpp:114]: Ah got it, thanks

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
I think the bond timeout won't quite work the way it is expected. Otherwise, LGTM and then we just need to make those updates to the website's configuration pages + migration guide

**Comment by SteveMacenski** [comment]
https://app.circleci.com/pipelines/github/ros-planning/navigation2/4267/workflows/30a56677-4cc3-46d3-8b6a-279a66e1a762/jobs/16101 also linting

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 14]
```diff
@@ -169,27 +170,6 @@ LifecycleManager::createBondConnection(const std::string & node_name)
   return true;
 }
 
-bool
-LifecycleManager::changeStateForNode(const std::string & node_name, std::uint8_t transition)
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp]: Why did you delete this function?

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 15]
```diff
@@ -144,18 +146,17 @@ LifecycleManager::destroyLifecycleServiceClients()
...
     bond_map_[node_name] =
       std::make_shared<bond::Bond>("bond", node_name, bond_client_node_);
     bond_map_[node_name]->setHeartbeatTimeout(timeout_s);
     bond_map_[node_name]->setHeartbeatPeriod(0.10);
     bond_map_[node_name]->start();
-    if (
-      !bond_map_[node_name]->waitUntilFormed(
-        rclcpp::Duration(rclcpp::Duration::from_seconds(timeout_s / 2))))
-    {
+    if (!bond_map_[node_name]->waitUntilFormed(rclcpp::Duration(timeout_ns / 2))) {
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp]: This API is deprecated, you cannot use Duration without a conversion or chrono literal

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 16]
```diff
@@ -44,7 +44,9 @@ LifecycleManager::LifecycleManager()
 
   node_names_ = get_parameter("node_names").as_string_array();
   get_parameter("autostart", autostart_);
-  bond_timeout_ = std::chrono::duration<double>(get_parameter("bond_timeout").as_double());
+  int bond_timeout_s;	  
+  get_parameter("bond_timeout", bond_timeout_s);
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:48]: bond_timeout_s cannot be an int, it must be a double.

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 17]
```diff
@@ -44,7 +44,9 @@ LifecycleManager::LifecycleManager()
 
   node_names_ = get_parameter("node_names").as_string_array();
   get_parameter("autostart", autostart_);
-  bond_timeout_ = std::chrono::duration<double>(get_parameter("bond_timeout").as_double());
+  int bond_timeout_s;	  
+  get_parameter("bond_timeout", bond_timeout_s);	
+  bond_timeout_ = std::chrono::milliseconds(bond_timeout_s * 1000);
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:49]: Please use chrono conversions, not just multiplying by 1000

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp (hunk 16 above)
Comment (SteveMacenski) [reply nav2_lifecycle_manager/src/lifecycle_manager.cpp:48]: Sorry, ROS2 is just really _really_ sensitive to parameter types

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp (hunk 17 above)
Comment (SteveMacenski) [reply nav2_lifecycle_manager/src/lifecycle_manager.cpp:49]: I know you're probably new to all the chrono things, I'm still learning some of them myself. This is a good opportunity to familiarize yourself with it :-)

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 18]
```diff
@@ -194,6 +192,7 @@ LifecycleManager::changeStateForNode(const std::string & node_name, std::uint8_t
   return true;
 }
 
+
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:195]: Please remove extra line here added.

File: doc/parameters/param_list.md (hunk 10 above)
Comment (SteveMacenski) [inline doc/parameters/param_list.md:429]: https://app.circleci.com/pipelines/github/ros-planning/navigation2/4296/workflows/381da9d4-f80f-46bf-9e61-bf2ad95590f0/jobs/16186 check the CI test jobs, you have some linting errors. You can reproduce these locally with `ament_cpplint` and `ament_uncrustify`

File: nav2_lifecycle_manager/src/lifecycle_manager.cpp [hunk 19]
```diff
@@ -348,9 +347,9 @@ LifecycleManager::checkBondConnections()
       // if one is down, bring them all down
       RCLCPP_ERROR(
         get_logger(),
-        "CRITICAL FAILURE: SERVER %s IS DOWN after not receiving a heartbeat for %i ms."
+        "CRITICAL FAILURE: SERVER %s IS DOWN after not receiving a heartbeat for %f s."
         " Shutting down related nodes.",
-        node_name.c_str(), static_cast<int>(bond_timeout_.count()));
+        node_name.c_str(), bond_timeout_.count());
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/src/lifecycle_manager.cpp:352]: I believe this is ms again.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
A few more bugs still, I'd recommend looking at the diff in github view and trying to track your changes to make sure units are consistent.

**Comment by simutisernestas** [comment]
should be ready now, i've rebased on main and updated docs accordingly

File: nav2_lifecycle_manager/test/launch_lifecycle_test.py [hunk 20]
```diff
@@ -32,7 +32,7 @@ def generate_launch_description():
             output='screen',
             parameters=[{'use_sim_time': False},
                         {'autostart': False},
-                        {'bond_timeout_ms': 0},
+                        {'bond_timeout': 0},
```
Comment (SteveMacenski) [inline nav2_lifecycle_manager/test/launch_lifecycle_test.py:35]: I think this needs to be `0.0` now (might be why the CI test isn't returning a result for the lifecycle tests)

**Review by SteveMacenski — COMMENTED** [review]
I see a CI failure that I think this 1 issue causes

**Comment by SteveMacenski** [comment]
Awesome! Thanks for the contribution and iterating with me!
